apps/prairielearn/src/ee/pages/instructorAiGenerateDraftEditor/RichTextEditor/index.tsx

```diff
@@ -0,0 +1,153 @@
+import { Strike } from '@tiptap/extension-strike';
+import { Subscript } from '@tiptap/extension-subscript';
+import { Superscript } from '@tiptap/extension-superscript';
+import { Text } from '@tiptap/extension-text';
+import { Underline } from '@tiptap/extension-underline';
+import { Focus, Selection, UndoRedo } from '@tiptap/extensions';
+import { EditorContent, useEditor } from '@tiptap/react';
+// import { BubbleMenu, FloatingMenu } from '@tiptap/react/menus';
+import { useState } from 'preact/compat';
+import prettierHtmlPlugin from 'prettier/plugins/html';
+import prettier from 'prettier/standalone';
+import { Card, Form } from 'react-bootstrap';
+
+import { PlQuestionPanel } from './extensions/pl-question-panel.js';
+import { RawHtml } from './extensions/raw-html.js';
+
+function formatHtmlWithPrettier(html: string): Promise<string> {
+  return prettier.format(html, {
+    parser: 'html',
+    plugins: [prettierHtmlPlugin],
+    printWidth: 100,
+    tabWidth: 2,
+  });
+}
+
+/**
+ * The main rich text editor component.
+ * @param params
+ * @param params.htmlContents - The initial HTML contents of the editor.
+ * @param params.csrfToken
+ */
+const RichTextEditor = ({
+  csrfToken: _csrfToken,
+  htmlContents,
+}: {
+  htmlContents: string | null;
+  csrfToken: string;
+}) => {
+  const editor = useEditor({
+    editorProps: {
+      attributes: {
+        'aria-label': 'Main content area, start typing to enter text.',
+        autocapitalize: 'off',
+        autocomplete: 'off',
+        autocorrect: 'off',
+      },
+    },
+    emitContentError: true,
+    enableContentCheck: true,
+    immediatelyRender: false,
+    onContentError: (event) => {
+      throw new Error(event.error.message);
+    },
+    onCreate: async ({ editor }) => {
+      const formattedHtml = await formatHtmlWithPrettier(editor.getHTML());
+      setFormattedHtml(formattedHtml);
+    },
+    onUpdate: async ({ editor }) => {
+      const formattedHtml = await formatHtmlWithPrettier(editor.getHTML());
+      setFormattedHtml(formattedHtml);
+    },
+    content: htmlContents,
+    extensions: [
+      ...new Set([
+        Blockquote,
+        Bold,
+        BulletList,
+        Code,
+        Document,
+        Dropcursor,
+        Focus,
+        Gapcursor,
+        Heading,
+        Italic,
+        ListItem,
+        ListKeymap,
+        OrderedList,
+        Paragraph,
+        PlQuestionPanel,
+        Selection,
+        Strike,
+        Subscript,
+        Superscript,
+        Text,
+        Underline,
+        UndoRedo,
+        // For this to work, we need all the nodes to be matched correctly by the other extensions.
+        RawHtml,
+      ]),
+    ],
+  });
+  const [formattedHtml, setFormattedHtml] = useState<string | null>(null);
+  const [debugMode, setDebugMode] = useState<boolean>(false);
+
+  if (htmlContents === null) {
+    return null;
+  }
+
+  if (editor === null) {
+    return null;
+  }
```

Comment:
I trust that this is just a weird TypeScript thing, but when on earth would `useEditor` return `null`? 🙃

Reply:
If initialization fails.